builtin/union_dataset.cc

```diff
@@ -0,0 +1,381 @@
+          "datasets/UnionDataset.md.html");
+
+extern std::shared_ptr<Dataset> (*createUnionDatasetFn) (MldbServer *, vector<std::shared_ptr<Dataset> > datasets);
+
+std::shared_ptr<Dataset> createUnionDataset(MldbServer * server, vector<std::shared_ptr<Dataset> > datasets)
+{
+    return std::make_shared<UnionDataset>(server, datasets);
+}
+
+struct UnionDataset::Itl
+    : public MatrixView, public ColumnIndex {
+
+    // Datasets that it was constructed with
+    vector<std::shared_ptr<Dataset> > datasets;
+
+    Itl(MldbServer * server, vector<std::shared_ptr<Dataset> > datasets) {
+        if (datasets.empty()) {
+            throw ML::Exception("Attempt to unify no datasets together");
+        }
+        this->datasets = datasets;
+    }
+
+    struct UnionRowStream : public RowStream {
+
+        UnionRowStream(const UnionDataset::Itl* source) : source(source)
+        {
+        }
+
+        virtual std::shared_ptr<RowStream> clone() const
+        {
+            return make_shared<UnionRowStream>(source);
+        }
+
+        /* set where the stream should start*/
+        virtual void initAt(size_t start)
+        {
+            throw ML::Exception("Unimplemented __FILE__ : __LINE__ ");
+        }
+
+        virtual RowName next()
+        {
+            uint64_t hash = (*it).first;
+            ++it;
+
+            return source->getRowName(RowHash(hash));
+        }
+
+        virtual const RowName & rowName(RowName & storage) const
+        {
+            uint64_t hash = (*it).first;
+            return storage = source->getRowName(RowHash(hash));
+        }
+
+        const UnionDataset::Itl* source;
+        IdHashes::const_iterator it;
+
+    };
+
+    pair<string, string> getStrFix(int pos, int length) const
+    {
+        auto fct = [] (int count, bool begin) -> string {
+            if (count == 0) {
+                return begin ? "[" : "]";
+            }
+            string s = "";
+            for (; count > 0; --count) {
+                s += "[]-";
+            }
+            return begin ? s + "[" : "]-" + s.substr(0, s.size() - 1);
+        };
+        return make_pair(fct(pos, true), fct(length - pos - 1, false));
+    }
+
+    virtual vector<Path>
+    getRowNames(ssize_t start = 0, ssize_t limit = -1) const
+    {
+        // Row names are []-[]-[]-...[] where the original row name goes into
+        // the [] at the index of the dataset within the union.
+        vector<Path> result;
+        for (int i = 0; i < datasets.size(); ++i) {
+            const auto & d = datasets[i];
+            string prefix;
+            string suffix;
+            tie(prefix, suffix) = getStrFix(i, datasets.size());
+            for (const auto & name: d->getMatrixView()->getRowNames()) {
+                result.push_back(Path(prefix + name.toUtf8String().rawString() + suffix));
+            }
+
+        }
+        return result;
+    }
+
+    virtual vector<RowHash>
+    getRowHashes(ssize_t start = 0, ssize_t limit = -1) const
```

Comment:
If I remove it it complains about the "pure virtual". Should I remove from the code base?